.aws/src/iam.ts

```diff
@@ -9,6 +9,114 @@ import { Construct } from 'constructs';
+  constructor(
+    scope: Construct,
+    name: string,
+    ecr_repo: ApplicationECR,
+    caller: DataAwsCallerIdentity
+  ) {
+    super(scope, name);
+    const orgId = config.OIDCOrgId;
+    const OIDCProviderId = `oidc.circleci.com/org/${orgId}`;
+    const OIDCProvider = new DataAwsIamOpenidConnectProvider(
+      this,
+      'OIDCProvider',
+      {
+        url: `https://${OIDCProviderId}`
+      }
+    );
+    const trustPolicy = new DataAwsIamPolicyDocument(
+      this,
+      'DataFlowsOIDCTrust',
+      {
+        version: '2012-10-17',
+        statement: [
+          {
+            effect: 'Allow',
+            actions: ['sts:AssumeRoleWithWebIdentity', 'sts:TagSession'],
+            principals: [
+              {
+                type: 'Federated',
+                identifiers: [OIDCProvider.arn]
+              }
+            ],
+            condition: [
+              {
+                test: 'StringEquals',
+                variable: `${OIDCProviderId}:aud`,
+                values: [orgId]
+              },
+              {
+                test: 'ForAnyValue:StringLike',
+                variable: `${OIDCProviderId}:sub`,
+                values: [`org/${orgId}/project/${config.OIDCProjectId}/user/*`]
+              }
+            ]
+          }
+        ]
+      }
+    );
+    const accessPolicy = new DataAwsIamPolicyDocument(
+      this,
+      'DataFlowsOIDCAccess',
+      {
+        version: '2012-10-17',
+        statement: [
+          {
+            effect: 'Allow',
+            actions: [
+              'ecr:BatchCheckLayerAvailability',
+              'ecr:GetRepositoryPolicy',
+              'ecr:DescribeRepositories',
+              'ecr:ListImages',
+              'ecr:DescribeImages',
+              'ecr:BatchGetImage',
+              'ecr:InitiateLayerUpload',
+              'ecr:UploadLayerPart',
+              'ecr:CompleteLayerUpload',
+              'ecr:PutImage'
+            ],
+            resources: [ecr_repo.repo.arn]
+          },
+          {
+            effect: 'Allow',
+            actions: [
+              'ecr:GetAuthorizationToken',
+              'ecs:RegisterTaskDefinition',
+              'ecs:ListTaskDefinitions',
+              'ecs:DescribeTaskDefinition',
+              'ecs:DeregisterTaskDefinition'
+            ],
+            resources: ['*']
+          },
+          {
+            effect: 'Allow',
+            actions: ['iam:PassRole'],
+            resources: [
+              `arn:aws:iam::${caller.accountId}:role/prefect-*`,
+              `arn:aws:iam::${caller.accountId}:role/data-flows-*`
+            ]
+          }
+        ]
+      }
+    );
+    new IamRole(this, 'DataFlowsOIDCRole', {
+      name: 'data-flows-circleci-oidc-role',
+      assumeRolePolicy: trustPolicy.json,
+      inlinePolicy: [
+        {
+          name: 'data-flows-circleci-oidc-policy',
+          policy: accessPolicy.json
+        }
+      ]
+    });
+  }
+}
```

Comment:
This is for creating an Open ID Connect based IAM role with least privilege for CircleCI to use.